File: Coding_Questions/C++/LongestPalindromicSubstring.cpp
// ---Longest Palindromic Substring---


bool is_palindromic(std::string str) {
    std::string str_original = str;
    std::reverse(str.begin(), str.end());
    return str_original == str;
}



std::string longest_palindromic_substring(std::string str) {
    std::string longest_palindrome(1, str[0]);

    for (size_t i = 0; i < str.length(); ++i) {
        for (size_t r = i; r < str.length(); ++r) {
            if (is_palindromic(str.substr(i, r + 1)) and str.substr(i, r + 1).length() > longest_palindrome.length()) {
                longest_palindrome = str.substr(i, r + 1);
            }
        }
    }

    return longest_palindrome;
}

File: Coding_Questions/Python/LongestPalindromicSubstring.py
# ---Longest Palindromic Substring---


def longest_palindromic_substring(string: str) -> str:
    longest_palindrome = "a"

    for i in range(len(string)):
        for r in range(i + 1, len(string)):
            if is_palindromic(string[i:r + 1]) and len(string[i:r + 1]) > len(
                    longest_palindrome
            ):
                longest_palindrome = string[i: r + 1]

    return longest_palindrome


def is_palindromic(string: str) -> bool:
    return string == "".join(reversed(list(string)))


print(longest_palindromic_substring("abaxyzzyxf"))  # "xyzzyx"
assert longest_palindromic_substring("abaxyzzyxf") == "xyzzyx"
